### PR TITLE
[WIP] API for probing a new path.

### PR DESCRIPTION
**Pull request by masa-koz**
## Description

I'm trying to add an API for probing a new path (QUIC_PARAM_CONN_ADD_LOCAL_ADDRESS).

## Testing

I've added a new test ("Basic/WithProbePathArgs.ProbePath/*").

## Documentation

N/A

### CODE REVIEW

File: src/core/binding.h [hunk 1]
```diff
@@ -225,6 +225,11 @@ typedef struct QUIC_BINDING {
     //
     CXPLAT_LIST_ENTRY Listeners;
 
+    //
+    // The connections registered on this binding.
+    //
+    CXPLAT_LIST_ENTRY Connections;
```
Comment (nibanks) [inline src/core/binding.h]: This shouldn't be on the binding itself, but as a part of the `Lookup`. Additionally, for performance, we don't want to use a list; that's why the Lookup uses hashtables. You're on the right path, but the lookup hash tables and entries will have to be modified to get this stuff to work.

Comment (masa-koz) [reply src/core/binding.h]: Thank you for your review! Is it acceptable to change the hashtable to register not the source CID itself but entry which contains the pointer to that source CID? If we change it, we could register a source CID into multiple bindings.

Comment (masa-koz) [reply src/core/binding.h]: I've implemented multiple hash entries for a Source CID.

File: src/core/lookup.c [hunk 2]
```diff
@@ -870,17 +905,22 @@ QuicLookupRemoveLocalCids(
...
                 Link);
-        if (CID->CID.IsInLookupTable) {
-            QuicLookupRemoveLocalCidInt(Lookup, CID);
+        while (CID->HashEntries.Next != NULL) {
+            QUIC_CID_HASH_ENTRY *CID1 =
+                CXPLAT_CONTAINING_RECORD(
+                    CxPlatListPopEntry(&CID->HashEntries),
+                    QUIC_CID_HASH_ENTRY,
+                    Link);
+            QuicLookupRemoveLocalCidInt(&CID1->Binding->Lookup, CID1);
```
Comment (masa-koz) [inline src/core/lookup.c]: Acquiring lock for this lookup is missing. But this function should be removed?

File: src/core/connection.h [hunk 3]
```diff
@@ -641,6 +641,11 @@ typedef struct QUIC_CONNECTION {
 
 } QUIC_CONNECTION;
 
+typedef struct QUIC_CONNECTION_LIST_ENTRY {
+    CXPLAT_LIST_ENTRY Link;
+    QUIC_CONNECTION *Connection;
+} QUIC_CONNECTION_LIST_ENTRY;
+
```
Comment (masa-koz) [inline src/core/connection.h]: This struct not necessary.

File: src/core/connection.c [hunk 4]
```diff
@@ -896,12 +920,32 @@ QuicConnGenerateNewSourceCid(
...
-                sizeof(QUIC_CID_HASH_ENTRY) + MsQuicLib.CidTotalLength);
+                sizeof(QUIC_CID_SLIST_ENTRY) + MsQuicLib.CidTotalLength);
             QuicConnFatalError(Connection, QUIC_STATUS_INTERNAL_ERROR, NULL);
             return NULL;
         }
-        if (!QuicBindingAddSourceConnectionID(Connection->Paths[0].Binding, SourceCid)) {
-            CXPLAT_FREE(SourceCid, QUIC_POOL_CIDHASH);
+
+        BOOLEAN Collision = FALSE;
+
```
Comment (masa-koz) [inline src/core/connection.c]: Acquiring MsQuicLib.DatapathLock is missing.

File: src/core/binding.c [hunk 5]
```diff
@@ -560,46 +561,101 @@ _IRQL_requires_max_(DISPATCH_LEVEL)
...
+        while (*Link1 != NULL) {
+            QUIC_CID_HASH_ENTRY* Entry1 = 
+                CXPLAT_CONTAINING_RECORD(
+                    *Link1,
+                    QUIC_CID_HASH_ENTRY,
+                    Link);
+            if (Entry1->Binding == Binding) {
+                QuicBindingRemoveSourceConnectionID(Binding, Entry1);
+                *Link1 = (*Link1)->Next;
+                CxPlatListPushEntry(&EntriesToFree, &Entry1->Link);
```
Comment (masa-koz) [inline src/core/binding.c:634]: Can't you free it here?

File: src/core/connection.c [hunk 6]
```diff
@@ -944,6 +994,12 @@ QuicConnGenerateNewSourceCid(
         SourceCid->Link.Next = NULL;
     }
 
+    for (uint8_t i = 0; i < Connection->PathsCount; ++i) {
+        if (Connection->Paths[i].Binding != NULL) {
+            QuicBindingAddSourceConnectionID(Connection->Paths[i].Binding, SourceCid);
+        }
+    }
```
Comment (masa-koz) [inline src/core/connection.c]: Shouldn't you move this to the insider of lock?

File: src/core/connection.c [hunk 7]
```diff
@@ -6564,6 +6675,153 @@ QuicConnParamSet(
         return QUIC_STATUS_SUCCESS;
     }
 
+    case QUIC_PARAM_CONN_ADD_LOCAL_ADDRESS: {
```
Comment (masa-koz) [inline src/core/connection.c:6680]: NEW_LOCAL_ADDRESS instead？

File: src/core/connection.c [hunk 8]
```diff
@@ -6564,6 +6675,153 @@ QuicConnParamSet(
...
+            break;
+        }
+
+        for (uint8_t i = 0; i < Connection->PathsCount; ++i) {
+            if (QuicAddrCompare(
+                &Connection->Paths[i].Route.LocalAddress,
+                LocalAddress)) {
+                break;
+            }
+        }
```
Comment (masa-koz) [inline src/core/connection.c:6718]: Check whether the local address already contains.

File: src/core/connection.c [hunk 9]
```diff
@@ -896,12 +920,45 @@ QuicConnGenerateNewSourceCid(
...
+            }    
+        }
+
+        if (!Collision) {
+            for (uint8_t i = 0; i < Connection->PathsCount; ++i) {
+                if (Connection->Paths[i].Binding != NULL) {
+                    QuicBindingAddSourceConnectionID(Connection->Paths[i].Binding, SourceCid);
+                }
+            }
+        }
```
Comment (masa-koz) [inline src/core/connection.c:956]: Move to outer of loop.

**Pull request closed without merging**